thewalrus/loop_hafnian_batch_gamma.py

```diff
@@ -0,0 +1,270 @@
+):  # pragma: no cover
+    r"""Calculate the loop hafnian batch for paired modes with threshould detectors.
+
+    Args:
+        A (array): input matrix.
+        D (array): vector to find loop hafnian batch.
+        fixed_edge_reps (array): fixed number of edge repetition.
+        batch_max (array): maximum number of photons for m mode.
+        odd_cutoff (array): cutoff for unpaired modes.
+        glynn (boolean): determines the method used to evaluate the loop hafnian batch.
+
+    Returns:
+        H_batch (array): matrix that contains batched loop hafnian with threshold detectors.
+    """
+    oddloop = D[:, 0]
+    oddV = A[0, :]
+
+    n = A.shape[0]
+    N_fixed = 2 * fixed_edge_reps.sum()  # number of photons
+
+    N_max = N_fixed + 2 * batch_max + odd_cutoff
+
+    edge_reps = np.concatenate((np.array([batch_max]), fixed_edge_reps))
+    steps = np.prod(edge_reps + 1)
+    # precompute binomial coefficients
+    max_binom = edge_reps.max() + odd_cutoff
+    binoms = precompute_binoms(max_binom)
+    n_D = D.shape[0]
+
+    H_batch = np.zeros((n_D, 2 * batch_max + odd_cutoff + 1), dtype=np.complex128)
+
+    for j in prange(steps):
+
+        Hnew = np.zeros((n_D, 2 * batch_max + odd_cutoff + 1), dtype=np.complex128)
+
+        kept_edges = find_kept_edges(j, edge_reps)
+        edges_sum = kept_edges.sum()
+
+        binom_prod = 1.0
+        for i in range(1, n // 2):
+            binom_prod *= binoms[edge_reps[i], kept_edges[i]]
+
+        if glynn:
+            delta = 2 * kept_edges - edge_reps
+        else:
+            delta = kept_edges
+
+        AX_S, XD_S, D_S, oddVX_S = get_submatrices(delta, A, D[0, :], oddV)
+
+        AX_S_copy = AX_S.copy()
+
+        for k in range(n_D):
+            XD_S, D_S = get_Dsubmatrices(delta, D[k, :])
+
+            f_even = f_loop(AX_S_copy, AX_S, XD_S, D_S, N_max)
+            f_odd = f_loop_odd(AX_S_copy, AX_S, XD_S, D_S, N_max, oddloop[k], oddVX_S)
+
+            for N_det in range(2 * kept_edges[0], 2 * batch_max + odd_cutoff + 1):
+                N = N_fixed + N_det
+                plus_minus = (-1.0) ** (N // 2 - edges_sum)
+
+                n_det_binom_prod = binoms[N_det // 2, kept_edges[0]] * binom_prod
+
+                if N_det % 2 == 0:
+                    Hnew[k, N_det] += n_det_binom_prod * plus_minus * f_even[N // 2]
+                else:
+                    Hnew[k, N_det] += n_det_binom_prod * plus_minus * f_odd[N]
+        H_batch += Hnew
+
+    if glynn:
+        for j in range(H_batch.shape[1]):
+            x = N_fixed + j
+            H_batch[:, j] *= 0.5 ** (x // 2)
+
+    return H_batch
+
+
+# pylint: disable = too-many-arguments, not-an-iterable
+@numba.jit(nopython=True, cache=True, parallel=True)
+def _calc_loop_hafnian_batch_gamma_odd(
+    A, D, fixed_edge_reps, batch_max, even_cutoff, glynn=True
+):  # pragma: no cover
+    r"""Calculate the loop hafnian batch for unpaired modes with threshould detectors.
```

Comment:
```suggestion
    r"""Calculate the loop hafnian batch for unpaired modes with threshold detectors.
```

Reply:
I think having a bit of clarification on how `_calc_loop_hafnian_batch_gamma_odd` and ` _calc_loop_hafnian_batch_gamma_even` are different could have some value. Currently the docstrings of both functions are exactly the same.